util/topsql/reporter/reporter.go

```diff
@@ -0,0 +1,461 @@
+}
+
+type planBinaryDecodeFunc func(string) (string, error)
+
+// RemoteTopSQLReporter implements a TopSQL reporter that sends data to a remote agent
+// This should be called periodically to collect TopSQL resource usage metrics
+type RemoteTopSQLReporter struct {
+	client ReportClient
+	// topSQLMap maps `sqlDigest-planDigest` to TopSQLDataPoints
+	topSQLMap map[string]*topSQLDataPoints
+
+	// normalizedSQLMap is an map, whose keys are SQL digest strings and values are normalized SQL strings
+	normalizedSQLMap atomic.Value // sync.Map
+
+	// normalizedPlanMap is a plan version of normalizedSQLMap
+	// this should only be set from the dedicated worker
+	normalizedPlanMap atomic.Value // sync.Map
+
+	// calling this can take a while, so should not block critical paths
+	planBinaryDecoder planBinaryDecodeFunc
+
+	collectCPUTimeChan chan *topSQLCPUTimeInput
+	reportDataChan     chan reportData
+
+	quit chan struct{}
+}
+
+// NewRemoteTopSQLReporter creates a new TopSQL struct
+//
+// planBinaryDecoder is a decoding function which will be called asynchronously to decode the plan binary to string
+// MaxStatementsNum is the maximum SQL and plan number, which will restrict the memory usage of the internal LFU cache
+func NewRemoteTopSQLReporter(client ReportClient, planDecoder planBinaryDecodeFunc) *RemoteTopSQLReporter {
+	tsr := &RemoteTopSQLReporter{
+		client:             client,
+		topSQLMap:          make(map[string]*topSQLDataPoints),
+		planBinaryDecoder:  planDecoder,
+		collectCPUTimeChan: make(chan *topSQLCPUTimeInput, collectCPUTimeChanLen),
+		reportDataChan:     make(chan reportData, 1),
+		quit:               make(chan struct{}),
+	}
+	tsr.normalizedSQLMap.Store(&sync.Map{})
+	tsr.normalizedPlanMap.Store(&sync.Map{})
+
+	go tsr.collectWorker()
+
+	go tsr.reportWorker()
+
+	return tsr
+}
+
+// Collect will drop the records when the collect channel is full
+// TODO: test the dropping behavior
+func (tsr *RemoteTopSQLReporter) Collect(timestamp uint64, records []tracecpu.SQLCPUTimeRecord) {
+	select {
+	case tsr.collectCPUTimeChan <- &topSQLCPUTimeInput{
+		timestamp: timestamp,
+		records:   records,
+	}:
+	default:
+		// ignore if chan blocked
+	}
+}
+
+func (tsr *RemoteTopSQLReporter) collectWorker() {
+	defer util.Recover("top-sql", "collectWorker", nil, false)
+
+	interval := variable.TopSQLVariable.ReportIntervalSeconds.Load()
+	reportTicker := time.NewTicker(time.Second * time.Duration(interval))
+	for {
+		select {
+		case input := <-tsr.collectCPUTimeChan:
+			tsr.collect(input.timestamp, input.records)
+		case <-reportTicker.C:
+			// Update report ticker if report interval changed.
+			if newInterval := variable.TopSQLVariable.ReportIntervalSeconds.Load(); newInterval != interval {
+				interval = newInterval
+				reportTicker.Reset(time.Second * time.Duration(interval))
+			}
+			tsr.sendToReport()
+		case <-tsr.quit:
+			return
+		}
+	}
+}
+
+func encodeKey(sqlDigest, planDigest []byte) string {
+	var buffer bytes.Buffer
```

Comment:
Use a pool and returning the string (and put back the buffer to the pool) is better.